# Number
In this notebook we will study Python hello world, then the numbers.

# Prerequisites

- Assume you have a Windows laptop with internet connection.
- You need install Visual Studio Code from Microsoft Store.
- Open a folder, such as Pyhton in your machine and create a notebooks folder under it.
- Copy any hello world ipynb file in the notebooks folder.
- Launch VSCode and open this folder.
- Install latest Python + Jupyter notebook when click install kernel.
- Set default environment from Get Started with Python Development.

## Hello World
The following python code will first print a line of text "Hello World"
Then we will assign this sentence to a variable and print it again.

In [6]:
# Assign "hello world" to a variable and print it again.
greeting_message = "Hello World!"
print(greeting_message)

Hello World!


#### Observation
From the above example, we observe following points:
1. Python does not use any statement end mark (for example ; in C/C++), end of line means end of statement.
2. We do not need to declare the type of a variable in advance (but we can do it, we will discuss later).

#### Variable Naming Convention 
We use lower case as name of variable, if the variable contains two english words, we use underscore '_' to connect them.
We should avoid using keyword as variable names. For user varialbles and function, we should also avoid starting with '_', because it is reserved by python internal.


### Comments
In python we can add comments lines by a leading '#'. The text after '#' is considered as explanation of the code and will not be executed.

## Numbers
Let's start some numeric value in python. Python is not a strong type language, so the numeric value is a mix of integer (32 bits), long (64 bits), float, double. We have a way to round up the float to integer.

The numberic value in python generally does not have limit, so it can be very big and you do not need to worry about overflow.

In [2]:
# Let's work on some basic arithmetic
a = 2 + 3
print(a)

a = 3 - 2
print(a)

a = 2 * 3
print(a)

a = 5 / 2
print('5 / 2 = ' + str(a))
print('int(5 / 2) = ' + str(int(a)))

a = 3**2
print('3**2' + str(a))

a = 2 * (4-2)
print(a)

a = 0.1 + 0.1
print(a)

a = 0.2 + 0.1
print(a)

large = 14_000_000_000_000_000_000_000_000_000_000_000
print("large = " , large)
googol = 10**100
print("googol = ", googol)

x, y, z = 1, 2, 3
print("x, y, z = ", x,y,z)

# The following is a constant value
MAX_CONNECTIONS = 5000
print('MAX_CONNECTIONS', MAX_CONNECTIONS)

# calculate module
print('7 % 3 = ', 7%3)

# scientific number
print('1.23456E+3 = ', 1.23456E+3)

# complex number
print('complex(3,5)', complex(3,5))


5
1
6
5 / 2 = 2.5
int(5 / 2) = 2
3**29
4
0.2
0.30000000000000004
large =  14000000000000000000000000000000000
googol =  10000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
x, y, z =  1 2 3
MAX_CONNECTIONS 5000
7 % 3 =  1
1.23456E+3 =  1234.56
complex(3,5) (3+5j)


#### Observation 
1. In Python there is no difference between integer and float. By default, the division result will be treated as float. This is quite different from other language, for example in C/C++ or Java, 5/2 = 2, not 2.5, but in python it is 2.5
2. Like all other languages, float will encounter precision issue, for example like 0.1 + 0.2. We will study precision round up later.
3. Python can use under score '_' to separate multiple digits in a number.
4. Python does not have capacity for a number, unlike C/C++ or Java, which will have integer(32 bits) and long (64 bits). A numeric value in python can be very big. But it is also not a double value, so you should not worry about precision of the large number.
5. We can assign multiple values to multiple variables in one statement.
6. Constant value can be represented by all upper case variables.
7. We can convert a float to integer by int().
8. We can convert numeric to string by str()

### Math Operations


| Operation | Result | Notes | 
| --- | --- | --- |
| x + y | sum of x and y |  |
| x - y | difference of x and y | |
| x * y | product of x and y | |
| x / y | quotient (商) of x and y | |
| x // y | floored quotient of x and y | (1) |
| x % y | remainder of x / y | (2) |
| -x | x negated | | 
| +x | x unchanged | |
| abs(x) | absolute value or magnitude of x | |
| int(x) | x converted to integer | (3)(6) | 
| float(x) | x converted to floating point | (4)(6) |
| complex(re, im) | a complex number with real part re, imaginary part im. im defaults to zero. | (6) |
| c.conjugate() | conjugate of the complex number c | |
| divmod(x, y) | the pair (x // y, x % y) | (2) |
| pow(x, y) | x to the power y | (5) |
| x ** y | x to the power y | (5) |

Notes:

1. Also referred to as integer division. The resultant value is a whole integer, though the result’s type is not necessarily int. The result is always rounded towards minus infinity: 1//2 is 0, (-1)//2 is -1, 1//(-2) is -1, and (-1)//(-2) is 0.

2. Not for complex numbers. Instead convert to floats using abs() if appropriate.

3. Conversion from float to int truncates, discarding the fractional part. See functions math.floor() and math.ceil() for alternative conversions.

4. float also accepts the strings “nan” and “inf” with an optional prefix “+” or “-” for Not a Number (NaN) and positive or negative infinity.

5. Python defines pow(0, 0) and 0 ** 0 to be 1, as is common for programming languages.

6. The numeric literals accepted include the digits 0 to 9 or any Unicode equivalent (code points with the Nd property).

See https://www.unicode.org/Public/14.0.0/ucd/extracted/DerivedNumericType.txt for a complete list of code points with the Nd property.

All numbers.Real types (int and float) also include the following operations:

| Operation | Result |
| --- | --- |
|  math.trunc(x)  |  x truncated to Integral  |
|  round(x[, n])  |  x rounded to n digits, rounding half to even. If n is omitted, it defaults to 0. |
|  math.floor(x)  |  the greatest Integral <= x  |
|  math.ceil(x)  |  the least Integral >= x  |

For additional numeric operations see the math and cmath modules.

In [3]:
import math

# division to get float result
print('7 / 2 = ' + str(7/2))

# division to get integer result only
print('7 // 2 = ' + str(7 // 2))

# get mod value
print('100 % 3 = ' + str(100 % 3))

# get quotient and remainder
(a, b) = divmod(100, 3)
print('divmod(100, 3) = ' + str(a) + ',' + str(b))

# calculate a^b
print("2^3 = ", 2**3)
print("3^2 = ", pow(3,2))

# get round up
a = 8 / 3
print("round(8/3, 2) = ", round(a, 2))



7 / 2 = 3.5
7 // 2 = 3
100 % 3 = 1
divmod(100, 3) = 33,1
2^3 =  8
3^2 =  9
round(8/3, 2) =  2.67


### Bitwise Operations on Integer Types

For any integer, internally we can convert to bits (in fact float can also be converted to bits, but less obvious)
Bitwise operations only make sense for integers. 

The result of bitwise operations is calculated as though carried out in two’s complement with an infinite number of sign bits.

The priorities of the binary bitwise operations are all lower than the numeric operations and higher than the comparisons; the unary operation ~ has the same priority as the other unary numeric operations (+ and -).

This table lists the bitwise operations sorted in ascending priority:

| Operation | Result | Notes |
| --- | --- | --- |
| x \| y | bitwise or of x and y | (4) |
| x ^ y | bitwise exclusive or of x and y | (4) |
| x & y | bitwise and of x and y | (4) |
| x << n | x shifted left by n bits | (1)(2) |
| x >> n | x shifted right by n bits | (1)(3) |
| ~x | the bits of x inverted |  |

Notes:
1. Negative shift counts are illegal and cause a ValueError to be raised.
2. A left shift by n bits is equivalent to multiplication by pow(2, n).
3. A right shift by n bits is equivalent to floor division by pow(2, n).
4. Remeber Python number does not have scope, so if we invert the bits, we should mask it with scope.


In [4]:
# shift bits right
print('9 >> 1 = ' + str(9 >> 1))
print('9 << 1 = ' + str(9 << 1))

# bit wise or
print('9 | 4 = ' + str(9 | 4))

# bit wise and
print('9 & 4 = ' + str(9 & 4))

# bit wise xor
print('9 ^ 4 = ' + str(9 ^ 4))

# invert a 4 bits integer
print('~9 = ', ~9 & 15)

# print an integer in bits
print("bin(9) =", bin(9))
print("bin((~9) & 255) =", bin((~9) & 15))
print("hex(9) =", hex(9))
print("hex(12) =", hex(12))

9 >> 1 = 4
9 << 1 = 18
9 | 4 = 13
9 & 4 = 0
9 ^ 4 = 13
~9 =  6
bin(9) = 0b1001
bin((~9) & 255) = 0b110
hex(9) = 0x9
hex(12) = 0xc


#### Observation 
1. For 9, bit wise it is 1001 and for 4 bit wise it is 0100.
2. We do not have bit flip ~x, this is because we need to specify unsigned integer.

### Assignment with calculations

Like C++ and Java, Python also support combining assignment and calculation together. Please look at following example

In [5]:
a = 8
print("a=8")
# a+=b
a+=3
print("a+=3:", a)
a = 8
a-=3
print("a-=3:", a)
a = 8
a*=3
print("a*=3:", a)
a = 8
a/=3
print("a/=3:", a)
a = 8
a%=3
print("a%=3:", a)
a = 8
a//=3
print("a//=3:", a)
a = 8
a**=3
print("a**=3:", a)

a=8
a+=3: 11
a-=3: 5
a*=3: 24
a/=3: 2.6666666666666665
a%=3: 2
a//=3: 2
a**=3: 512


### Round up float
We can round up float with specified decimal position

In [5]:
# round up pi
pi = 3.1415926
print(round(pi, 3))

minus_pi = -3.1415926
print(round(minus_pi, 3))

# round up by hundreds or thousands
a = 1627731
print(round(a, -2))
print(round(a, -3))


3.142
-3.142
1627700
1628000


### Format Numbers for Output

In [17]:
# format float
x = 1234.56789
print("x =", x)

# two decimal places of accuracy
print("format(x, '0.2f')", format(x, '0.2f'))

#right justified in 10 chars, one-digit accuracy
print("format(x, '>10.1f')", format(x, '>10.1f'))

#left justified
print("format(x, '<10.1f')", format(x, '<10.1f'))

#center justified
print("format(x, '^10.1f')", format(x, '^10.1f'))

#center justified
print("format(x, ',')", format(x, ','))
print("format(x, '0,.1f')", format(x, '0,.1f'))

x = 1234.56789
format(x, '0.2f') 1234.57
format(x, '>10.1f')     1234.6
format(x, '<10.1f') 1234.6    
format(x, '^10.1f')   1234.6  
format(x, ',') 1,234.56789
format(x, '0,.1f') 1,234.6


### Binary, Octal and Hexadecimal

In [20]:
x = 1234
print("bin(x)", bin(x))
print("oct(x)", oct(x))
print("hex(x)", hex(x))
print("format(x, 'b')", format(x, 'b'))
print("format(x, 'o')", format(x, 'o'))
print("format(x, 'x')", format(x, 'x'))
print("hex(x)", hex(x))
print("int('4d2', 16)", int('4d2', 16))
print("int('10011010010', 2)", int('10011010010', 2))



bin(x) 0b10011010010
oct(x) 0o2322
hex(x) 0x4d2
format(x, 'b') 10011010010
format(x, 'o') 2322
format(x, 'x') 4d2
hex(x) 0x4d2
int('4d2', 16) 1234
int('10011010010', 2) 1234


### Complex

In [22]:
a = complex(2, 4)
b = 3-5j
print("a = ", a, "; b= ", b)
print("a + b = ", a + b)
print("a.real = ", a.real, "; a.imag = ", a.imag, "; a.conjugate() = ", a.conjugate())

# import cmath
import cmath
print("cmath.sqrt(-2) = ", cmath.sqrt(-2))

a =  (2+4j) ; b=  (3-5j)
a + b =  (5-1j)
a.real =  2.0 ; a.imag =  4.0 ; a.conjugate() =  (2-4j)
cmath.sqrt(-2) =  1.4142135623730951j


#### Infinity and NaNs
inf is infinity and nan is not a number

In [ ]:
a = float('inf')
b = float('-inf')
c = float('nan')

## Exercise
1.1 If we borrow 50,000 dollar from the bank with a fixed annual interest rate as 3%, with term of 5 years. How much I will pay at the end of 5 years.




1.2 If I buy a car with 20000 dollar, and if the value of car drops 15% in the first 3 years, how much the value left at the end of 3 years?



1.3 Assume you have an hourly salary rate as $20 and if you work 8 hours per day, and you work 300 days in a year. and you have an monthly expense $2000, how much money you can save per year.



1.4 The speed of sound is 340 m /s. We design a rockt as same speed of sound. The distance from the earth to the moon is 384,400 km. Can you calculate how many days and how many hours does it take for the rocket travel from earth to the moon?


1.5 Assume you have two points in a plane, (x1, y1) and (x2, y2). Calculate the distance of these two points.


1.6 We have two measurements on the temperature, Celsisus, C and Farenheit F. The 0 degree in Celsisus is equal to 32 degree Farenheit. The coversion formula is as below:
Celsisus = (Farenheit - 32) * 5 / 9
Farenheit = Celsisus * (9 /5) + 32


1.7 You get a loan of 1000,000 $ from the bank and the interest rate is 5%, with the fixed term of 30 years. Can you calculate monthlt payment? The rule is that you have to pay off the interest generated every year and pay off the loan in the 30 years, you should pay the same amount every month.
You can use the following formula:
Monthly Payment = (loan * month rate) / (1 - 1/(1+ month rate) ^ (years * 12))

1.8 calculate the sum of 1+3+5+...+999
